clients/drcachesim/tools/filter/encoding_filter.h

```diff
@@ -0,0 +1,173 @@
+ * DAMAGE.
+ */
+
+#ifndef _ENCODING_FILTER_H_
+#define _ENCODING_FILTER_H_ 1
+
+#include "record_filter.h"
+#include "trace_entry.h"
+#include "utils.h"
+
+#include <cstring>
+#include <vector>
+
+/* We are not exporting the defines in core/ir/isa_regdeps/encoding_common.h, so we
+ * redefine DR_ISA_REGDEPS alignment requirement here.
+ */
+#define REGDEPS_ALIGN_BYTES 4
+
+#define REGDEPS_MAX_ENCODING_LENGTH 16
+
+namespace dynamorio {
+namespace drmemtrace {
+
+class encoding_filter_t : public record_filter_t::record_filter_func_t {
+public:
+    encoding_filter_t()
+    {
+    }
+
+    void *
+    parallel_shard_init(memtrace_stream_t *shard_stream,
+                        bool partial_trace_filter) override
+    {
+        dcontext_.dcontext = dr_standalone_init();
+        return nullptr;
+    }
+
+    bool
+    parallel_shard_filter(trace_entry_t &entry, void *shard_data,
+                          std::vector<trace_entry_t> &last_encoding) override
+    {
+        /* TODO i#6662: modify trace_entry_t header entry to regdeps ISA, instead of the
+         * real ISA of the incoming trace.
+         */
+
+        /* We have encoding to convert.
+         * Normally the sequence of trace_entry_t(s) looks like:
+         * [encoding,]+ instr_with_PC, [read | write]*
+         * (+ = one or more, * = zero or more)
+         * If we enter here, trace_entry_t is instr_with_PC.
+         */
+        if (is_any_instr_type(static_cast<trace_type_t>(entry.type)) &&
+            !last_encoding.empty()) {
+            /* Gather real ISA encoding bytes looping through all previously saved
+             * encoding bytes in last_encoding.
+             */
+            const app_pc pc = reinterpret_cast<app_pc>(entry.addr);
+            byte encoding[MAX_ENCODING_LENGTH];
+            memset(encoding, 0, sizeof(encoding));
+            uint encoding_offset = 0;
+            for (auto &trace_encoding : last_encoding) {
+                memcpy(encoding + encoding_offset, trace_encoding.encoding,
+                       trace_encoding.size);
+                encoding_offset += trace_encoding.size;
+            }
+
+            /* Genenerate the real ISA instr_t by decoding the encoding bytes.
+             */
+            instr_t instr;
+            instr_init(dcontext_.dcontext, &instr);
+            app_pc next_pc = decode_from_copy(dcontext_.dcontext, encoding, pc, &instr);
+            if (next_pc == NULL || !instr_valid(&instr)) {
+                instr_free(dcontext_.dcontext, &instr);
+                error_string_ =
+                    "Failed to decode instruction " + to_hex_string(entry.addr);
+                return false;
+            }
+
+            /* Convert the real ISA instr_t into a regdeps ISA instr_t.
+             */
+            instr_t instr_regdeps;
+            instr_init(dcontext_.dcontext, &instr_regdeps);
+            instr_convert_to_isa_regdeps(dcontext_.dcontext, &instr, &instr_regdeps);
```

Comment:
Does this not return a success value to check here?

Reply:
It does not, but the encoding to DR_ISA_REGDEPS (which uses the output of `instr_convert_to_isa_regdeps()`) does, so checking that.